### ISSUE TITLE
Very low level discrepancies (3 reads in ~88 billion primary reads)

### ISSUE THREAD

**Issue by keiranmraine** [bwa-mem2/bwa-mem2#109]
The issue evaluates 96 samples of WGS data totalling 44,497,901,587 read pairs (read length = 151).

* High level overview
  * Headers
  * Mapping stats
  * Read vs read comparison
* Discrepant samples
  * XA tag difference
  * Different primary mapping location

## High level overview

High level checks of mapping info, headers and overall reproduction of `bwa mem` with `bwa-mem2`.

### Headers

This shows that the PG entry of the PG line has been switched to `bwa-mem2` as appropriate:

```bash
$ cat sample.lst | xargs -I {} bash -c "diff -sq <(samtools view -H 2451/{}/mapped_sample/{}.sample.dupmarked.bam | grep -cP 'PN:bwa\s') <(samtools view -H 2463/{}/mapped_sample/{}.sample.dupmarked.bam | grep -c 'PN:bwa-mem2')" | grep -v 'are identical'
<no output>
```

No output is the expected result.

### Mapping stats

We use a tool called `bam_stats` ([code][pcap-core-c]) to calculate the following values for each mapped file (some excluded from listing).  The data is collated by read-group in the file, one line for each.

```plain
#_mapped_bases
#_mapped_bases_r1
#_mapped_bases_r2
#_divergent_bases
#_divergent_bases_r1
#_divergent_bases_r2
#_total_reads
#_total_reads_r1
#_total_reads_r2
#_mapped_reads
#_mapped_reads_r1
#_mapped_reads_r2
#_mapped_reads_properly_paired
#_gc_bases_r1
#_gc_bases_r2
mean_insert_size
insert_size_sd
median_insert_size
#_duplicate_reads
#_mapped_pairs
#_inter_chr_pairs
#_qc_fail_r1
#_qc_fail_r2
```

The data were compared via:

```bash
$ cat sample.lst | xargs -I {} diff -qs 2451/{}/mapped_sample/{}.sample.dupmarked.bam.bas 2463/{}/mapped_sample/{}.sample.dupmarked.bam.bas | grep -v 'are identical'
<no output>
```

All data shows as equal.

### Read vs read comparison

Next we compare each bam file using `comm` to check the content of each read mapping is exact between the 2 sets of 96 samples.

The underlying command used for comparison was:

```bash
comm -23 --nocheck-order <(samtools view 2451/{}/mapped_sample/{}.sample.dupmarked.bam) <(samtools view 2463/{}/mapped_sample/{}.sample.dupmarked.bam)
```

On completion all logs showing any output were evaluated.  This identified *2/96 samples as being discrepant*.  The changes are described in the next section

## Discrepant samples

Samples numbers as per rank in sorted list.

### XA tag difference

Affects sample 32.

Deep in a single XA tag with 46 alternatives one entry is less fragmented

MD: `41S33M7D16M15I3M5I31M7S` -> `41S33M7D16M23I31M7S`

```bash
# bwa mem
A00537:0167:BHLC2HDSXX:3:1228:25310:17268	161	chr14	67910412	0	7S40M104S	=	37750595	-30159666	CGGGCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATTTAAATGGTTTATATATTACAAAAAAAAAAAAAAAATTATTTTTAAAAAAAAAAAAAAAAAAAAAAAAATAAAAAAAAATATAAAAAAAAAAAAAAAAAAAAAATAAAAA	F,,,,,:,,,FFFFFFFFFFFFF:,:,FFFFFFF::,,,,,F:,FF,,,,:F,F,,,,,,,,FFFFF,FFFFFFFF:,,,F:,:,:,,FFFFFFFFFF,FF,F:,F,,,,,,,:,,,,,,,,,:,,::FFFFFFF,,,:::FF:,,:,F,F	NM:i:0	MD:Z:40AS:i:40	XS:i:37	SA:Z:chr3,81768948,+,97S17M1D37M,0,3;chr15,43418176,+,80S31M40S,0,0;	XA:Z:chr17,+77598918,4S37M110S,0;chr3,+198089704,4S37M110S,0;chr1,-160484241,110S37M4S,0;chr1,-42709140,110S37M4S,0;chr21,+39554936,5S36M110S,0;chrX,+86901285,6S36M109S,0;chr15,-24137345,108S36M7S,0;chr6,-23300139,108S36M7S,0;chr11,-33614582,110S36M5S,0;chr10,+102106537,2S36M113S,0;chr12,-77514883,109S36M6S,0;chr19,+32318180,7S36M108S,0;chr11,-14638583,110S36M5S,0;chr7,-76846891,108S36M7S,0;chr5,+61056888,5S36M110S,0;chr5,+7673238,5S36M110S,0;chr8,-80405620,110S36M5S,0;chr17,+38939292,5S36M110S,0;chr18,+43693303,7S35M109S,0;chr5,+112048085,10S35M106S,0;chr4,+57282152,7S35M109S,0;chr1,+198028168,7S35M109S,0;chr4,-132750411,109S35M7S,0;chr6,+91866408,7S35M109S,0;chr10,+5951143,7S35M109S,0;chr8,-73378347,109S35M7S,0;chr19,-44546238,109S35M7S,0;chr9,-26847195,109S35M7S,0;chr12,-13098462,113S35M3S,0;chr11,-21326051,113S35M3S,0;chr2,+142870377,7S35M109S,0;chr20,+55432880,4S35M112S,0;chr12,+22944917,7S35M109S,0;chr4,+31405351,7S35M109S,0;chr7,-141575655,115S36M,1;chr2,-156220606,108S34M9S,0;chr15,-51191888,107S34M10S,0;chr22,+10537100,9S34M108S,0;chr13,+51641826,9S34M108S,0;chr22,+15708514,9S34M108S,0;chr2,-117826035,116S33M2S,0;chr7,-107770157,41S33M7D16M15I3M5I31M7S,28;chr13,-28293388,115S33M3S,0;chr10,-87534037,107S33M11S,0;chr19,-4947491,115S33M3S,0;chr3_KI270784v1_alt,-142213,110S37M4S,0;chr18_KI270864v1_alt,-85677,109S35M7S,0;	MQ:i:60	MC:Z:151M	ms:i:5064	RG:Z:465202
# bwa-mem2
A00537:0167:BHLC2HDSXX:3:1228:25310:17268	161	chr14	67910412	0	7S40M104S	=	37750595	-30159666	CGGGCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATTTAAATGGTTTATATATTACAAAAAAAAAAAAAAAATTATTTTTAAAAAAAAAAAAAAAAAAAAAAAAATAAAAAAAAATATAAAAAAAAAAAAAAAAAAAAAATAAAAA	F,,,,,:,,,FFFFFFFFFFFFF:,:,FFFFFFF::,,,,,F:,FF,,,,:F,F,,,,,,,,FFFFF,FFFFFFFF:,,,F:,:,:,,FFFFFFFFFF,FF,F:,F,,,,,,,:,,,,,,,,,:,,::FFFFFFF,,,:::FF:,,:,F,F	NM:i:0	MD:Z:40AS:i:40	XS:i:37	SA:Z:chr3,81768948,+,97S17M1D37M,0,3;chr15,43418176,+,80S31M40S,0,0;	XA:Z:chr17,+77598918,4S37M110S,0;chr3,+198089704,4S37M110S,0;chr1,-160484241,110S37M4S,0;chr1,-42709140,110S37M4S,0;chr21,+39554936,5S36M110S,0;chrX,+86901285,6S36M109S,0;chr15,-24137345,108S36M7S,0;chr6,-23300139,108S36M7S,0;chr11,-33614582,110S36M5S,0;chr10,+102106537,2S36M113S,0;chr12,-77514883,109S36M6S,0;chr19,+32318180,7S36M108S,0;chr11,-14638583,110S36M5S,0;chr7,-76846891,108S36M7S,0;chr5,+61056888,5S36M110S,0;chr5,+7673238,5S36M110S,0;chr8,-80405620,110S36M5S,0;chr17,+38939292,5S36M110S,0;chr18,+43693303,7S35M109S,0;chr5,+112048085,10S35M106S,0;chr4,+57282152,7S35M109S,0;chr1,+198028168,7S35M109S,0;chr4,-132750411,109S35M7S,0;chr6,+91866408,7S35M109S,0;chr10,+5951143,7S35M109S,0;chr8,-73378347,109S35M7S,0;chr19,-44546238,109S35M7S,0;chr9,-26847195,109S35M7S,0;chr12,-13098462,113S35M3S,0;chr11,-21326051,113S35M3S,0;chr2,+142870377,7S35M109S,0;chr20,+55432880,4S35M112S,0;chr12,+22944917,7S35M109S,0;chr4,+31405351,7S35M109S,0;chr7,-141575655,115S36M,1;chr2,-156220606,108S34M9S,0;chr15,-51191888,107S34M10S,0;chr22,+10537100,9S34M108S,0;chr13,+51641826,9S34M108S,0;chr22,+15708514,9S34M108S,0;chr2,-117826035,116S33M2S,0;chr7,-107770157,41S33M7D16M23I31M7S,31;chr13,-28293388,115S33M3S,0;chr10,-87534037,107S33M11S,0;chr19,-4947491,115S33M3S,0;chr3_KI270784v1_alt,-142213,110S37M4S,0;chr18_KI270864v1_alt,-85677,109S35M7S,0;	MQ:i:60	MC:Z:151M	ms:i:5064	RG:Z:465202
```

### Different primary mapping location

Affects sample 81.

55 reads are initially indicated as discrepant by comm, however this is due to the file going out of sync.

The issue is actually a single MAPQ=0 readpair being shifted:

```bash
# bwa
HX2_27119:7:2216:3772:5634	99	chr9	60579935	0	151M	=	60580328	544	TAGTTTTTATGGGAAGATATTTCCTTTTTCAACATAGGCCTCAAAGCGCTCCAAATGTCTACTTCCAGGTAGTGTAGAAAGAGTGTTTCAAACCTGCTCTATAAAAGGGAATATTCAACTCTGTGACTTGAATGCAAACATCACAAAGCAC	<-AAAFJJJJJJJFFJJFJJAJJJJJJJJJJJJJFFFJJAJJJJ<JFJJ<AFJJJJFJJJ7F<-<<<<J7<<-F-<JFJFJ<JA-AJFJFFAFJ<AFJ-A7AAJJFJJAFJJJJJJJJJFJFFFJFJFFJJ<JFJJ-77----AAJAAFJJ	NM:i:1	MD:Z:74C76	AS:i:146	XS:i:146	MQ:i:0	MC:Z:151M	ms:i:3026	RG:Z:401295
HX2_27119:7:2216:3772:5634	659	chr9	60580328	0	151M	=	60579935	-544	TAATATCCACTTGCAGATTCTACAGAAAGAGTGCTTCAAAACTGCTCTGTCCAAAGGTTTTTTTGACTCTGTTACTTGAGTACTCACCTCACATTCCCTTTTCTGAGATTGCTTCTGTCTGGTTTTTAGGGGAAGATATTTCCTTTTTCAA	-F-7-<7--FAA77---<-AF<-F-FF77F7<77FA-<777A7FJFFFA<FF-7--7--7-<7-7A-<F-F7-F7---<7--7-7----F-<-------A-7F-<-<--<---------F-JJJJJJF-F-JJJJJJJJJJJJJJJFFFAA	NM:i:19	MD:Z:0A23A32A0A0G0G2C0A18A3A5A0G0G0A0A0G9A12C8A20	AS:i:60	XS:i:65	MQ:i:0	MC:Z:151M	ms:i:5167	mm:A:Y	RG:Z:401295
# bwa-mem2
HX2_27119:7:2216:3772:5634	99	chr9	60610723	0	151M	=	60611116	544	TAGTTTTTATGGGAAGATATTTCCTTTTTCAACATAGGCCTCAAAGCGCTCCAAATGTCTACTTCCAGGTAGTGTAGAAAGAGTGTTTCAAACCTGCTCTATAAAAGGGAATATTCAACTCTGTGACTTGAATGCAAACATCACAAAGCAC	<-AAAFJJJJJJJFFJJFJJAJJJJJJJJJJJJJFFFJJAJJJJ<JFJJ<AFJJJJFJJJ7F<-<<<<J7<<-F-<JFJFJ<JA-AJFJFFAFJ<AFJ-A7AAJJFJJAFJJJJJJJJJFJFFFJFJFFJJ<JFJJ-77----AAJAAFJJ	NM:i:1	MD:Z:74C76	AS:i:146	XS:i:146	MQ:i:0	MC:Z:151M	ms:i:3026	RG:Z:401295
HX2_27119:7:2216:3772:5634	659	chr9	60611116	0	151M	=	60610723	-544	TAATATCCACTTGCAGATTCTACAGAAAGAGTGCTTCAAAACTGCTCTGTCCAAAGGTTTTTTTGACTCTGTTACTTGAGTACTCACCTCACATTCCCTTTTCTGAGATTGCTTCTGTCTGGTTTTTAGGGGAAGATATTTCCTTTTTCAA	-F-7-<7--FAA77---<-AF<-F-FF77F7<77FA-<777A7FJFFFA<FF-7--7--7-<7-7A-<F-F7-F7---<7--7-7----F-<-------A-7F-<-<--<---------F-JJJJJJF-F-JJJJJJJJJJJJJJJFFFAA	NM:i:19	MD:Z:0A23A32A0A0G0G2C0A18A3A5A0G0G0A0A0G9A12C8A20	AS:i:60	XS:i:65	MQ:i:0	MC:Z:151M	ms:i:5167	mm:A:Y	RG:Z:401295
```

BLAT supports both read 1 locations:

```none
   ACTIONS      QUERY SCORE START   END QSIZE IDENTITY  CHROM                    STRAND  START       END   SPAN
---------------------------------------------------------------------------------------------------------------
browser details read1   149     1   151   151    99.4%  chr9                     +    60610723  60610873    151
browser details read1   149     1   151   151    99.4%  chr9                     +    60579935  60580085    151
```

BLAT also supports both of the read 2 locations, +1 as the first base is a mismatch in both bwa versions:

```none
   ACTIONS      QUERY SCORE START   END QSIZE IDENTITY  CHROM                    STRAND  START       END   SPAN
---------------------------------------------------------------------------------------------------------------
browser details read2   114     2   151   151    88.0%  chr9                     +    60580329  60580478    150
browser details read2   114     2   151   151    88.0%  chr9                     +    60611117  60611266    150
```

[pcap-core-c]: https://github.com/cancerit/PCAP-core/tree/develop/c

**Comment by keiranmraine** [comment]
Please note that both runs were with exactly the same number of threads.  `-K` was suggested as a possible solution but from what I can see both `bwa` and `bwa-mem2` use the same calculation and `chunk_size` when this is not provided.

**Comment by yuk12** [comment]
Thanks for the report. can you share the data on discrepant reads and the details of the runs: command line, reference?

**Comment by keiranmraine** [comment]
Quick update, I am working to reproduce on minimal dataset.

**Comment by keiranmraine** [comment]
For sample 32 - change to XA mapping.

I've managed to reproduce the difference with 3001 pairs.  The data has been emailed directly to @yuk12 & @sanchit-misra.

```
zcat 32_3k_up.fq.gz | $BWA mem -T 30 -Y -p -t 1 GRCh38_full_analysis_set_plus_decoy_hla/genome.fa - > ${BWA}_3k_up.sam
```

comm:

```
comm -23 --nocheck-order <(samtools view bwa-32_3k_up.sam) <(samtools view bwa-mem2-32_3k_up.sam)
A00537:0167:BHLC2HDSXX:3:1228:25310:17268	161	chr14	67910412	0	7S40M104S	=	37750595	-30159668	CGGGCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATTTAAATGGTTTATATATTACAAAAAAAAAAAAAAAATTATTTTTAAAAAAAAAAAAAAAAAAAAAAAAATAAAAAAAAATATAAAAAAAAAAAAAAAAAAAAAATAAAAA	F,,,,,:,,,FFFFFFFFFFFFF:,:,FFFFFFF::,,,,,F:,FF,,,,:F,F,,,,,,,,FFFFF,FFFFFFFF:,,,F:,:,:,,FFFFFFFFFF,FF,F:,F,,,,,,,:,,,,,,,,,:,,::FFFFFFF,,,:::FF:,,:,F,F	NM:i:0	MD:Z:40MC:Z:151M	AS:i:40	XS:i:37	SA:Z:chr3,81768948,+,97S17M1D37M,0,3;chr2,14912698,+,79S31M41S,0,0;	XA:Z:chr17,+77598918,4S37M110S,0;chr3,+198089704,4S37M110S,0;chr1,-160484241,110S37M4S,0;chr1,-42709140,110S37M4S,0;chr8,-80405620,110S36M5S,0;chr7,-76846891,108S36M7S,0;chr11,-14638583,110S36M5S,0;chr10,+102106537,2S36M113S,0;chr5,+7673238,5S36M110S,0;chr12,-77514883,109S36M6S,0;chr5,+61056888,5S36M110S,0;chrX,+86901285,6S36M109S,0;chr6,-23300139,108S36M7S,0;chr17,+38939292,5S36M110S,0;chr15,-24137345,108S36M7S,0;chr11,-33614582,110S36M5S,0;chr21,+39554936,5S36M110S,0;chr19,+32318180,7S36M108S,0;chr10,+5951143,7S35M109S,0;chr4,+31405351,7S35M109S,0;chr8,-73378347,109S35M7S,0;chr20,+55432880,4S35M112S,0;chr6,+91866408,7S35M109S,0;chr12,+22944917,7S35M109S,0;chr4,+57282152,7S35M109S,0;chr4,-132750411,109S35M7S,0;chr11,-21326051,113S35M3S,0;chr2,+142870377,7S35M109S,0;chr18,+43693303,7S35M109S,0;chr12,-13098462,113S35M3S,0;chr5,+112048085,10S35M106S,0;chr9,-26847195,109S35M7S,0;chr1,+198028168,7S35M109S,0;chr19,-44546238,109S35M7S,0;chr22,+15708514,9S34M108S,0;chr2,-156220606,108S34M9S,0;chr15,-51191888,107S34M10S,0;chr22,+10537100,9S34M108S,0;chr7,-141575655,115S36M,1;chr13,+51641826,9S34M108S,0;chr13,-28293388,115S33M3S,0;chr19,-4947491,115S33M3S,0;chr2,-117826035,116S33M2S,0;chr10,-87534037,107S33M11S,0;chr7,-107770157,41S33M7D16M15I3M5I31M7S,28;chr3_KI270784v1_alt,-142213,110S37M4S,0;chr18_KI270864v1_alt,-85677,109S35M7S,0;
```

Specific issue:

* `bwa`
  * `chr7,-107770157,41S33M7D16M15I3M5I31M7S,28`
* `bwa-mem2`
  * `chr7,-107770157,41S33M7D16M23I31M7S,31`

Trimming the first 1k reads causes the `bwa-mem2` output to match `bwa`.

**Comment by keiranmraine** [comment]
For sample 81 - different primary mapping location for pair

This one has been difficult to trim down to a small set of reads.  The problem is as follows:

* 1 thread - match bwa
* 10 threads - mismatch
* 10 threads -K 10000000 - match bwa

We need to understand why specifying `-K 10000000` changes the behaviour when that is the default value for both `bwa` and `bwa-mem2`

**Comment by yuk12** [comment]
Point 3 keeps the batch size smaller in comparison to point 2 where batch size is 10x more due to 10 threads.  Threads could be interfering with each other's memory (not sure), but then the discrepancy you are seeing could be different across the runs.
If the point2 data is not too big, share it I will narrow it down.

**Comment by keiranmraine** [comment]
I think I may have mixed up a file in the above comment, bwa-mem2 gives a different answer to bwa consistently with 10 threads.

> Point 3 keeps the batch size smaller in comparison to point 2 where batch size is 10x more due to 10 threads. Threads could be interfering with each other's memory (not sure), but then the discrepancy you are seeing could be different across the runs.

It looks like both `bwa` and `bwa-mem2` have a default value of `10000000` when the user doesn't define `-K`, so provided the thread value is the same the result should be the same.

https://github.com/lh3/bwa/blob/c4a010d19fb2e00f1dda560a64bc9d2d8d4a1de5/bwamem.c#L99

vs

https://github.com/bwa-mem2/bwa-mem2/blob/5122aa5943be34ae6c404886bedd4e2ce849cbbb/src/bwamem.cpp#L132

When `-K` is defined the calculation seems to have the same result:

https://github.com/bwa-mem2/bwa-mem2/blob/96b21e44b9a4829beb29a15a57c359d1fe9eaec1/src/fastmap.cpp#L940-L945

vs

https://github.com/lh3/bwa/blob/c4a010d19fb2e00f1dda560a64bc9d2d8d4a1de5/fastmap.c#L386

> If the point2 data is not too big, share it I will narrow it down.

In the original run (where this was identified) both `bwa` and `bwa-mem2` used 10 threads, I've trimmed the data and run both `bwa` and `bwa-mem2` with and without `-K 10000000`.  This shows that the change in result is consistent with 10 threads regarless of `-K` being set:

* `ID:bwa	PN:bwa	VN:0.7.17-r1188	CL:bwa mem -T 30 -Y -K 10000000 -p -t 10 GRCh38/genome.fa -`
  * Without -K = chr9:60579935, chr9:60580328
  * With -K = chr9:60579935, chr9:60580328
* `ID:bwa-mem2	PN:bwa-mem2	VN:2.0	CL:bwa-mem2 mem -T 30 -Y -K 10000000 -p -t 10 GRCh38/genome.fa -`
  * Without -K = chr9:60610723, chr9:60611116
  * With -K = chr9:60610723, chr9:60611116

bwa-mem2 version 2.1, code pulled via recursive git clone of tag, compiled with gcc.  I've emailed @yuk12 regarding transfering ~500MB of data which reproduces this issue.

**Comment by yuk12** [comment]
can you send me the exact command line used for both of them? I am not able to reproduce the issue.

**Comment by keiranmraine** [comment]
Both issues are based around this command, as shown in the PG lines 2 comments above.  In the case of the very small XA data only 1 thread is used.  In the second larger data set the `-K` option is added (as above).

```
cat reads_interleaved.fq | $BWA mem -T 30 -Y -p -t $CPUS GRCh38_full_analysis_set_plus_decoy_hla/genome.fa -
```

Please note that my binaries are standard GCC compilations.  I can redo to reconfirm including chipset and versions where I see the problem if you continue to be unable to reproduce.

**Comment by yuk12** [comment]
Update: I am able to reproduce the issue now. Working on fixing it.

**Comment by yuk12** [comment]
I've fixed the issue, it was a bug in BSW. Doing sanity performance checks now.

**Comment by yuk12** [comment]
The issue is resolved now. We have also made small optimizations to sam processing part (commit: e4f0276, thanks to @arun-sub) which gave us overall 1%-3% enhancements.

**Comment by keiranmraine** [comment]
I'll see if I can repeat the large scale test this side of the new year.

**Comment by yuk12** [comment]
Sounds good. Please verify the fix to this issue on your side before starting the large scale run.

**Comment by keiranmraine** [comment]
I can confirm that the test data I provided is now behaving as expected.  I will move to the full scale test (96 samples 88 billion pairs) to confirm there are no unexpected knock on effects.

**Comment by keiranmraine** [comment]
Our small CI has indicated a discrepancy in a different dataset following this change, I'm investigating now

**Comment by keiranmraine** [comment]
Clean CI run rather than a restart following build issue (wget fail) seems to have cleared that issue

**Comment by yuk12** [comment]
Great!!

**Comment by keiranmraine** [comment]
I'm sorry to say that the large scale run of 96 samples has identified issues in 3 different samples that were not present in 2.1 using commit id 9cd69b3f383a5cbb5e47b9e08e4c11392e72a510.

Do you want a new issue starting for this?

**Comment by yuk12** [comment]
Thanks for testing the code, Keiran. Can you check those three samples on the latest commit (``d19c7bb``)? this will verify whether the improvements we put in sam processing part are behind it or not (I've disabled the sam improvement code for now). If the issues persist then please let us know the details. We can continue on this thread. Thanks.

**Comment by keiranmraine** [comment]
It appears that `d19c7bb` resolves the first of the 3, I'll rerun the remainder but results will be after the break.

**Comment by yuk12** [comment]
Sure, sounds good.

**Comment by keiranmraine** [comment]
In all 3 cases d19c7bb resolves the issues seen in 9cd69b3.

I have small test data for each that can be shared directly as previously, but all require surrounding reads.

## Sample 65

Affects 1 readpair (`HX5_27176:6:2206:11140:13228`), due to change in supplementary alignment location, `chr1` -> `chr22_KI270736v1_random`.

```none
< HX5_27176:6:2206:11140:13228	2129	chrY	11291640	0	38M113S	=	56832296	45540620	CCACCCCACTCCACTCCACTCCACTCCACTCCACTCCAGTCCACTCCACTCCTCTCCATTAAATCCCAATCCATTGCATCCTACTCCATTCCATTCCATTCCTTTGCTTTCTTTCGATAGGATCTCACTTTGTCACACAGGGTGGAGTGCA	)7-)A)F-JAJFAJFJF-FFJJAJ<F7-JFJA-JFJJJJFJJAFJJJJJFAA<JJFJJJJJJJJJF<AFJJJJJJAJFJF<FJFAJJFJJJJAFJJJJJJJJFFFJFFAJJFJJJJJJJJFFJJJJJJJJJJJJF7AJJFJJJJFJFFAAA	NM:i:1	MD:Z:4T33	MC:Z:22S95M34S	AS:i:33	XS:i:33	SA:Z:chrY,10673032,-,20S131M,6,12;	XA:Z:chr15,+80153176,113S38M,1;chrY,-10990090,38M113S,1;chr2,+238125773,113S38M,1;chr22_KI270736v1_random,+137282,113S38M,1;chrY,-10654090,38M113S,1;chr7,+57491724,113S38M,1;chrY,-10744972,38M113S,1;chr7,+57491719,113S38M,1;chrY,-10654085,38M113S,1;chr2,+238125768,113S38M,1;chr1,+30781284,113S38M,1;chr12,+131588255,121S30M,0;chrY,-10862981,28M123S,0;chrY,-10677163,28M123S,1;chrUn_KN707959v1_decoy,+67,118S27M6S,0;
---
> HX5_27176:6:2206:11140:13228	2129	chrY	11291640	0	38M113S	=	56832296	45540620	CCACCCCACTCCACTCCACTCCACTCCACTCCACTCCAGTCCACTCCACTCCTCTCCATTAAATCCCAATCCATTGCATCCTACTCCATTCCATTCCATTCCTTTGCTTTCTTTCGATAGGATCTCACTTTGTCACACAGGGTGGAGTGCA	)7-)A)F-JAJFAJFJF-FFJJAJ<F7-JFJA-JFJJJJFJJAFJJJJJFAA<JJFJJJJJJJJJF<AFJJJJJJAJFJF<FJFAJJFJJJJAFJJJJJJJJFFFJFFAJJFJJJJJJJJFFJJJJJJJJJJJJF7AJJFJJJJFJFFAAA	NM:i:1	MD:Z:4T33	MC:Z:22S95M34S	AS:i:33	XS:i:33	SA:Z:chrY,10673032,-,20S131M,6,12;	XA:Z:chrY,-10990090,38M113S,1;chr7,+57491719,113S38M,1;chr22_KI270736v1_random,+137282,113S38M,1;chrY,-10654085,38M113S,1;chr1,+30781284,113S38M,1;chrY,-10744972,38M113S,1;chr15,+80153176,113S38M,1;chr7,+57491724,113S38M,1;chr2,+238125773,113S38M,1;chr2,+238125768,113S38M,1;chrY,-10654090,38M113S,1;chr12,+131588255,121S30M,0;chrY,-10862981,28M123S,0;chrY,-10677163,28M123S,1;chrUn_KN707959v1_decoy,+67,118S27M6S,0;
```

## Sample 78

One additional entry in the SA tag, 13th of 14 items in new mapping.

```none
$ diff <(samtools view mapped/2451/PD41779f_lo0008/mapped_sample/PD41779f_lo0008.sample.dupmarked.bam chrUn_KN707896v1_decoy:6765-6765) <(samtools view mapped/2463/PD41779f_lo0008/mapped_sample/PD41779f_lo0008.sample.dupmarked.bam chrUn_KN707896v1_decoy:6765-6765)
< HX7_27184:7:1111:18284:11294	2193	chrUn_KN707896v1_decoy	6765	0	142M9S	chr4	49143638	0	TCCATTCCATTCCCTTCCACTCGCCTCGATTCCATTCCATTCCATTCCATTCCATTCCATTCCATTCCATTCCGATCCGTTCCGTTCCATTCCATTCCATTCCATTCCATTCCGTTCCGTTCCATTCCATTCCATTCCATTCCATTCCATT	-<F7FJFA-<<J<7--)7)))7))-A---FJFA--77---AF7F7FFF7-<A7-77AF-<JFF-FFFFJJFF77---A---A7--AF7A-FF7J<AJA<JJ<<--JAFF-AFJJ7FJA-777AJFJJJJJJJJJJJJJJJJJJJJJFFFAA	NM:i:9	MD:Z:14C9G1T20T15C4T5T3A4A58	MC:Z:45M5I29M72S	AS:i:97	XS:i:94	SA:Z:chr22_KI270737v1_random,32889,-,36S82M5D33M,0,8;chrY,11300681,-,23M1I35M92S,0,5;	XA:Z:chr22_KI270737v1_random,-32889,36S82M5D33M,8;chr22,+16352791,33M5D82M36S,8;chr4,-49144452,151M,15;chrUn_KN707896v1_decoy,-26165,65M10I76M,17;chrUn_KN707896v1_decoy,-8650,147M4S,11;chrUn_KN707896v1_decoy,-27620,43S99M9S,2;chrUn_KN707896v1_decoy,-5912,65M10I76M,18;chrUn_KN707896v1_decoy,-21867,38S109M4S,5;chrUn_KN707896v1_decoy,-27337,119M5D32M,17;chrUn_KN707896v1_decoy,-24134,28M5I46M10I62M,21;chrUn_JTFH01001870v1_decoy,-1913,25S126M,9;chrUn_KN707896v1_decoy,-20084,48S94M9S,3;chrUn_KN707896v1_decoy,-5652,119M5D32M,18;	RG:Z:404176
---
> HX7_27184:7:1111:18284:11294	2193	chrUn_KN707896v1_decoy	6765	0	142M9S	chr4	49143638	0	TCCATTCCATTCCCTTCCACTCGCCTCGATTCCATTCCATTCCATTCCATTCCATTCCATTCCATTCCATTCCGATCCGTTCCGTTCCATTCCATTCCATTCCATTCCATTCCGTTCCGTTCCATTCCATTCCATTCCATTCCATTCCATT	-<F7FJFA-<<J<7--)7)))7))-A---FJFA--77---AF7F7FFF7-<A7-77AF-<JFF-FFFFJJFF77---A---A7--AF7A-FF7J<AJA<JJ<<--JAFF-AFJJ7FJA-777AJFJJJJJJJJJJJJJJJJJJJJJFFFAA	NM:i:9	MD:Z:14C9G1T20T15C4T5T3A4A58	MC:Z:45M5I29M72S	AS:i:97	XS:i:94	SA:Z:chr22_KI270737v1_random,32889,-,36S82M5D33M,0,8;chrY,11300681,-,23M1I35M92S,0,5;	XA:Z:chr22_KI270737v1_random,-32889,36S82M5D33M,8;chr22,+16352791,33M5D82M36S,8;chr4,-49144452,151M,15;chrUn_KN707896v1_decoy,-26165,65M10I76M,17;chrUn_KN707896v1_decoy,-8650,147M4S,11;chrUn_KN707896v1_decoy,-27620,43S99M9S,2;chrUn_KN707896v1_decoy,-5912,65M10I76M,18;chrUn_KN707896v1_decoy,-21867,38S109M4S,5;chrUn_KN707896v1_decoy,-27337,119M5D32M,17;chrUn_KN707896v1_decoy,-24134,28M5I46M10I62M,21;chrUn_JTFH01001870v1_decoy,-1913,25S126M,9;chrUn_KN707896v1_decoy,-20084,48S94M9S,3;chrUn_KN707896v1_decoy,-27620,48S103M,5;chrUn_KN707896v1_decoy,-5652,119M5D32M,18;	RG:Z:404176
```

## Sample 87

Order of SA changes:

```none
< 3:2147:7627:8406	2225	chr2	32110636	0	22S40M89S	chr20	55625108	0	TATATATTATATATAATATGGTGTATATATAGTATATATAATATGGTATATATAGTATATATGATATGGTATATATAGTATATATGATATGGTATATATGGTATATATATAGTATATATAGTATATATGTATACTATATATAATATATACT	FFFFFFFFFFFFFFFFFF:FFFFFFFF:FFFFFFFFFFFFFFF:FFF:FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NM:i:2	MD:Z:18G3A17	MC:Z:83S38M4I26M	AS:i:30	XS:i:30	SA:Z:chr8,38074904,+,9M4D12M2I29M1I12M4D21M3I16M46S,0,18;	XA:Z:chr12,-18098331,22S45M84S,3;chr5,+125615273,106S16M3D29M,5;chr9,-102363918,27S35M89S,2;chr17,-73669272,15S25M111S,0;chr8,+62870554,111S25M15S,0;chr6,+134687555,97S30M24S,1;chrUn_JTFH01001762v1_decoy,+5871,83S22M2D23M23S,4;	RG:Z:458520
---
> 3:2147:7627:8406	2225	chr2	32110636	0	22S40M89S	chr20	55625108	0	TATATATTATATATAATATGGTGTATATATAGTATATATAATATGGTATATATAGTATATATGATATGGTATATATAGTATATATGATATGGTATATATGGTATATATATAGTATATATAGTATATATGTATACTATATATAATATATACT	FFFFFFFFFFFFFFFFFF:FFFFFFFF:FFFFFFFFFFFFFFF:FFF:FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NM:i:2	MD:Z:18G3A17	MC:Z:83S38M4I26M	AS:i:30	XS:i:30	SA:Z:chr8,38074904,+,9M4D12M2I29M1I12M4D21M3I16M46S,0,18;	XA:Z:chr12,-18098331,22S45M84S,3;chr5,+125615273,106S16M3D29M,5;chr6,+134687555,97S30M24S,1;chr9,-102363918,27S35M89S,2;chr17,-73669272,15S25M111S,0;chr8,+62870554,111S25M15S,0;chrUn_JTFH01001762v1_decoy,+5871,83S22M2D23M23S,4;	RG:Z:458520
```

Fewer entries in SA:

```none
< 3:1478:32145:36871	577	chr6	80641258	0	31S23M12I20M3I8M2D11M3I22M3I15M	chr3	66726360	0	AGTATATATTATATATAGTATACATATATACTATATATACTATATATATACCATATATACCATATCATATATACTATATATACCATATCATATATACTATATATACCATATTATATATACTATATATACACCATATTATATATAATATATA	FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFF	NM:i:25	MD:Z:51^TA28T12C6	AS:i:38	XS:i:38	SA:Z:chr5,86420613,+,50M101S,0,4;	XA:Z:chr6,+80641236,31S23M12I8M2D12M3I19M3I22M3I15M,25;chr8,+38074904,9M4D12M2I29M1I12M4D21M3I16M46S,18;chr13,-87679828,92S41M18S,1;chr3,-66726353,21S40M2I14M74S,4;chr15,-65864457,7S17M1I7M7D23M7D23M7D21M52S,24;chrX,+4443578,21S38M92S,1;chr2,+132493004,23S35M3I22M68S,6;chr6,+80641344,31S25M10I15M3I13M2D11M3I25M15S,21;chr8,-85186656,68S18M1I41M23S,5;chrX,+114805815,31S32M3I8M2D14M63S,6;chr4,+152889719,2S21M1D10M5D11M1D32M3D10M3I22M4D18M22S,25;chrUn_JTFH01000770v1_decoy,-163,7M2D15M7D45M4I28M6I23M23S,28;chrUn_JTFH01000770v1_decoy,-163,23S31M9D36M4I15M6I13M1D23M,29;	MQ:i:0	MC:Z:28M2I19M102S	ms:i:5478	mm:A:Y	RG:Z:458520
---
> 3:1478:32145:36871	577	chr6	80641258	0	31S23M12I20M3I8M2D11M3I22M3I15M	chr3	66726360	0	AGTATATATTATATATAGTATACATATATACTATATATACTATATATATACCATATATACCATATCATATATACTATATATACCATATCATATATACTATATATACCATATTATATATACTATATATACACCATATTATATATAATATATA	FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFF	NM:i:25	MD:Z:51^TA28T12C6	AS:i:38	XS:i:38	SA:Z:chr5,86420613,+,50M101S,0,4;	XA:Z:chr6,+80641236,31S23M12I8M2D12M3I19M3I22M3I15M,25;chr8,+38074904,9M4D12M2I29M1I12M4D21M3I16M46S,18;chr13,-87679828,92S41M18S,1;chr3,-66726353,21S40M2I14M74S,4;chr15,-65864457,7S17M1I7M7D23M7D23M7D21M52S,24;chrX,+4443578,21S38M92S,1;chr2,+132493004,23S35M3I22M68S,6;chr8,-85186656,68S18M1I41M23S,5;chr6,+80641344,31S25M10I15M3I13M2D11M3I25M15S,21;chrX,+114805815,31S32M3I8M2D14M63S,6;chr4,+152889719,2S21M1D10M5D11M1D32M3D10M3I22M4D18M22S,25;chrUn_JTFH01000770v1_decoy,-163,7M2D15M7D45M4I28M6I23M23S,28;	MQ:i:0	MC:Z:28M2I19M102S	ms:i:5478	mm:A:Y	RG:Z:458520
```

**Comment by arun-sub** [comment]
Thanks Keiran. I can take a look at these discrepancies and try to reproduce it. Could you please share the minimal dataset if possible, or both the read and its corresponding mate for the differing alignments ?

This is your command line right ?
```cat reads_interleaved.fq | $BWA mem -T 30 -Y -p -t $CPUS GRCh38_full_analysis_set_plus_decoy_hla/genome.fa -```

**Comment by keiranmraine** [comment]
@arun-sub yes that command is equivalent, actually had bwa read from the file, also each of the above were reproduced with a single thread.

```
$BWA mem -T 30 -Y -p -t 1 GRCh38_full_analysis_set_plus_decoy_hla.fa $FQ_I > ${BWA}-${SAMP}.sam
```

**Comment by keiranmraine** [comment]
@arun-sub can you confirm you have received the data?

**Comment by arun-sub** [comment]
Yes, thank you. Will update here after reproducing the issue.

**Comment by yuk12** [comment]
@keiranmraine do you see any issues with the commit `d19c7bb` on the remaining query datasets? We can go ahead with this commit for the full testing and subsequent release. We can incorporate Arun's patch in the next round of features/enhancements.

**Comment by keiranmraine** [comment]
@yuk12 @arun-sub , I've just managed to confirm that the results from Arun's patch 7b7ae26 give the correct results for the 2 problem samples.  I can rerun the remaining overnight but will likely be later in the day (UK) tomorrow before the results are fully confirmed for those.

**Comment by keiranmraine** [comment]
Update for anyone tracking this.  We've isolated what appear to be the remaining issues to a subset of the binaries/modes.  Results from avx512bw and avx2 match legacy bwa.  avx, sse41 and sse42 have a very low level mapping discrepancy.

**Comment by sanchit-misra** [comment]
Update: the remaining issues have been fixed and verified by Keiran. No more discrepancies between bwa-mem and bwa-mem2 for the ~88 Billion reads that Keiran tested for.
